# Using ROBOKOP's Quick Question Answering Service

The expand, enrich, and similarity services each offer bite-sized approaches to interacting with ROBOKOP.  The ROBOKOP "quick" service offers a slightly more complex approach to answering questions.   We'll be using the following function to call the quick service.  Note that calling the service requires posting a question.  We'll discuss the format of the question below.

In [22]:
robokop_server='robokop.renci.org'

In [39]:
import os
import sys
module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)
from gg_functions import parse_answer

In [23]:
import requests

def quick(question,max_results=None,output_format=None,max_connectivity=None):
    url=f'http://{robokop_server}:80/api/simple/quick/'
    if max_results is not None:
        url += f'?max_results={max_results}'
    if output_format is not None:
        j = '&' if '?' in url else '?'
        url += f'{j}output_format={output_format}'
    if max_connectivity is not None:
        j = '&' if '?' in url else '?'
        url += f'{j}max_connectivity={max_connectivity}'
    response = requests.post(url,json=question)
    print( f"Return Status: {response.status_code}" )
    if response.status_code == 200:
        return response.json()
    return response

## Question Format and Basic Usage

The question is a python dictionary.  It takes a key `machine_question`, which is a dictionary containing a list of `nodes` and a list of `edges`.   Each node object needs an integer `id`.   Any node can also have a `curie` specifying that node.  The edges define connections between the nodes using the identifiers of the nodes as source and targets. The following function shows how to construct a single-hop question.  It takes a specified node of `type1` and looks for any node of `type2`.

In [24]:
def make_one_step_question(type1, id1, type2,rebuild = None):
    question = {
                'machine_question': {
                    'nodes': [
                        {
                            'id': 'n0',
                            'curie': id1,
                            'type': type1
                        },
                        {
                            'id': 'n1',
                            'type': type2
                        }
                    ],
                    'edges': [
                        {
                            'id': 'e0',
                            'source_id': 'n0',
                            'target_id': 'n1'
                        }
                    ]
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild'] = 'True'
    return question

Here, we will specify a one-hop question asking for the phenotypes associated with Fanconi Anemia (MONDO:0019391).  We first construct the question then use it to call the quick service.  In fact, this is how the expand function is implemented internally.  The output is returned in the KG standard answer format.

In [25]:
q = make_one_step_question('disease','MONDO:0019391','phenotypic_feature')
r = quick(q)

Return Status: 200


In [26]:
# Uncomment to see a big long answer
import json
#print( json.dumps(r, indent=4))

The output can be simplified with the function parse_answer.  That function is loaded into this notebook from the gg_functions library.

In [40]:
parse_answer(r)

e0 - edge_source      e0 - type    n1 - id   \
0    biolink.disease_get_phenotype  has_phenotype  HP:0001915   
1    biolink.disease_get_phenotype  has_phenotype  HP:0003002   
2    biolink.disease_get_phenotype  has_phenotype  HP:0004808   
3    biolink.disease_get_phenotype  has_phenotype  HP:0002860   
4    biolink.disease_get_phenotype  has_phenotype  HP:0100615   
5    biolink.disease_get_phenotype  has_phenotype  HP:0002488   
6    biolink.disease_get_phenotype  has_phenotype  HP:0002721   
7    biolink.disease_get_phenotype  has_phenotype  HP:0000135   
8    biolink.disease_get_phenotype  has_phenotype  HP:0004430   
9    biolink.disease_get_phenotype  has_phenotype  HP:0004810   
10   biolink.disease_get_phenotype  has_phenotype  HP:0006721   
11   biolink.disease_get_phenotype  has_phenotype  HP:0001908   
12   biolink.disease_get_phenotype  has_phenotype  HP:0001009   
13   biolink.disease_get_phenotype  has_phenotype  HP:0000824   
14   biolink.disease_get_phenotype  has_phenotype  HP:0001251   
15   biolink.disease_get_phenotype  has_phenotype  HP:0000789   
16   biolink.disease_get_phenotype  has_phenotype  HP:0010972   
17   biolink.disease_get_phenotype  has_phenotype  HP:0001402   
18   biolink.disease_get_phenotype  has_phenotype  HP:0100806   
19   biolink.disease_get_phenotype  has_phenotype  HP:0001972   
20   biolink.disease_get_phenotype  has_phenotype  HP:0000953   
21   biolink.disease_get_phenotype  has_phenotype  HP:0000286   
22   biolink.disease_get_phenotype  has_phenotype  HP:0003254   
23   biolink.disease_get_phenotype  has_phenotype  HP:0008064   
24   biolink.disease_get_phenotype  has_phenotype  HP:0005510   
25   biolink.disease_get_phenotype  has_phenotype  HP:0005505   
26   biolink.disease_get_phenotype  has_phenotype  HP:0001900   
27   biolink.disease_get_phenotype  has_phenotype  HP:0011018   
28   biolink.disease_get_phenotype  has_phenotype  HP:0004377   
29   biolink.disease_get_phenotype  has_phenotype  HP:0007354   
..                             ...            ...         ...   
220  biolink.phenotype_get_disease  has_phenotype  HP:0011842   
221  biolink.phenotype_get_disease  has_phenotype  HP:0030976   
222  biolink.phenotype_get_disease  has_phenotype  HP:0001928   
223  biolink.phenotype_get_disease  has_phenotype  HP:0410004   
224  biolink.phenotype_get_disease  has_phenotype  HP:0100737   
225  biolink.phenotype_get_disease  has_phenotype  HP:0011875   
226  biolink.phenotype_get_disease  has_phenotype  HP:0000202   
227  biolink.phenotype_get_disease  has_phenotype  HP:0011338   
228  biolink.phenotype_get_disease  has_phenotype  HP:0010989   
229  biolink.phenotype_get_disease  has_phenotype  HP:0000152   
230  biolink.phenotype_get_disease  has_phenotype  HP:0003256   
231  biolink.phenotype_get_disease  has_phenotype  HP:0000153   
232  biolink.disease_get_phenotype  has_phenotype  HP:0000268   
233  biolink.disease_get_phenotype  has_phenotype  HP:0006824   
234  biolink.disease_get_phenotype  has_phenotype  HP:0002245   
235  biolink.disease_get_phenotype  has_phenotype  HP:0001347   
236  biolink.disease_get_phenotype  has_phenotype  HP:0000340   
237  biolink.disease_get_phenotype  has_phenotype  HP:0001537   
238  biolink.disease_get_phenotype  has_phenotype  HP:0000582   
239  biolink.disease_get_phenotype  has_phenotype  HP:0100587   
240  biolink.disease_get_phenotype  has_phenotype  HP:0002414   
241  biolink.disease_get_phenotype  has_phenotype  HP:0000453   
242  biolink.disease_get_phenotype  has_phenotype  HP:0000027   
243  biolink.disease_get_phenotype  has_phenotype  HP:0001053   
244  biolink.disease_get_phenotype  has_phenotype  HP:0002823   
245  biolink.disease_get_phenotype  has_phenotype  HP:0005344   
246  biolink.disease_get_phenotype  has_phenotype  HP:0004349   
247  biolink.disease_get_phenotype  has_phenotype  HP:0008053   
248  biolink.disease_get_phenotype  has_phenotype  HP:0000813   
249  biolink.disease_get_phenotype  has_phenotype  HP

## Multi-step questions

It's straightforward to generalize this linear query to an N-item path.  This function constructs a question from a list of types and a list of identifiers.  The types are the types of nodes that will be traversed along the path, and the identifiers represent fixed elements in the path.  The length of types and ids should be equal, and free nodes should specify an id of None:

In [27]:
def make_N_step_question(types,curies,rebuild = None):
    question = {
                'machine_question': {
                    'nodes': [],
                    'edges': []
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild'] = 'True'
    ei = 0
    for i,t in enumerate(types):
        newnode = {'id': f'n{i}', 'type': t}
        if curies[i] is not None:
            newnode['curie'] = curies[i]
        question['machine_question']['nodes'].append(newnode)
        if i > 0:
            question['machine_question']['edges'].append( {'id': f'e{ei}','source_id': f'n{i-1}', 'target_id': f'n{i}'})
            ei += 1
    return question

We can recapitulate our previous question with this new function like this:

In [28]:
newq = make_N_step_question(['disease','phenotypic_feature'],['MONDO:0019391',None])
q == newq

True

Now we could expand to a longer query.  The following question will start at the disease MONDO:0019391, go to a gene, and from there to a biological process or activity.

In [29]:
two_step_question = make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,None])
print(json.dumps(two_step_question,indent=4))

{
    "machine_question": {
        "nodes": [
            {
                "id": "n0",
                "type": "disease",
                "curie": "MONDO:0019391"
            },
            {
                "id": "n1",
                "type": "gene"
            },
            {
                "id": "n2",
                "type": "biological_process_or_activity"
            }
        ],
        "edges": [
            {
                "id": "e0",
                "source_id": "n0",
                "target_id": "n1"
            },
            {
                "id": "e1",
                "source_id": "n1",
                "target_id": "n2"
            }
        ]
    }
}


In [30]:
two_step_answer = quick(two_step_question)

Return Status: 200


We can extract the node names along the paths with a simple function:

In [42]:
two_step_nodes = parse_answer(two_step_answer,node_list=['n0','n1','n2'],edge_list=[],node_properties=['name'])
two_step_nodes

n0 - name  n1 - name   \
0    Fanconi Anemia      FANCG   
1    Fanconi Anemia     RAD51C   
2    Fanconi Anemia      BRCA2   
3    Fanconi Anemia        TNF   
4    Fanconi Anemia      FANCA   
5    Fanconi Anemia      FANCL   
6    Fanconi Anemia      RAD51   
7    Fanconi Anemia     MAD2L2   
8    Fanconi Anemia        TNF   
9    Fanconi Anemia      RAD51   
10   Fanconi Anemia      BRCA1   
11   Fanconi Anemia      ERCC4   
12   Fanconi Anemia      BRCA1   
13   Fanconi Anemia        MX1   
14   Fanconi Anemia      FANCF   
15   Fanconi Anemia      FANCC   
16   Fanconi Anemia     RAD51C   
17   Fanconi Anemia      FANCG   
18   Fanconi Anemia      FANCA   
19   Fanconi Anemia       SLX4   
20   Fanconi Anemia      FANCL   
21   Fanconi Anemia      ERCC4   
22   Fanconi Anemia      BRIP1   
23   Fanconi Anemia      FANCC   
24   Fanconi Anemia      FANCG   
25   Fanconi Anemia       SLX4   
26   Fanconi Anemia      BRIP1   
27   Fanconi Anemia      BRIP1   
28   Fanconi Anemia      FANCA   
29   Fanconi Anemia     RAD51C   
..              ...        ...   
220  Fanconi Anemia        BLM   
221  Fanconi Anemia        WRN   
222  Fanconi Anemia      RFWD3   
223  Fanconi Anemia       REV1   
224  Fanconi Anemia      BRCA2   
225  Fanconi Anemia      BRCA2   
226  Fanconi Anemia      BRCA2   
227  Fanconi Anemia       DNA2   
228  Fanconi Anemia      BRCA2   
229  Fanconi Anemia      BRCA2   
230  Fanconi Anemia    L3MBTL1   
231  Fanconi Anemia       TERT   
232  Fanconi Anemia      ATRIP   
233  Fanconi Anemia       RMI1   
234  Fanconi Anemia      FANCE   
235  Fanconi Anemia    TP53BP1   
236  Fanconi Anemia       RNF8   
237  Fanconi Anemia        BLM   
238  Fanconi Anemia        WRN   
239  Fanconi Anemia        ATM   
240  Fanconi Anemia      DDX11   
241  Fanconi Anemia     FBXO18   
242  Fanconi Anemia       FAN1   
243  Fanconi Anemia      MUS81   
244  Fanconi Anemia    DCLRE1B   
245  Fanconi Anemia    FAAP100   
246  Fanconi Anemia       RNF8   
247  Fanconi Anemia    DCLRE1A   
248  Fanconi Anemia       POLN   
249  Fanconi Anemia      CENPS   

                                            n2 - name      score  
0                                response to radiation  0.558728  
1                                    DNA recombination  0.555804  
2                                          hemopoiesis  0.544295  
3               tumor necrosis factor receptor binding  0.523751  
4                     regulation of cell proliferation  0.493756  
5                     regulation of cell proliferation  0.489189  
6                                    DNA recombination  0.489088  
7                                        cell division  0.488393  
8                                inflammatory response  0.465679  
9                                           DNA repair  0.452533  
10                     regulation of apoptotic process  0.445331  
11                                      response to UV  0.445122  
12                    regulation of cell proliferation  0.444966  
13                              innate immune response  0.444392  
14                              protein ubiquitination  0.444250  
15                                          DNA repair  0.441425  
16                                          DNA repair  0.440971  
17                                          DNA repair  0.440748  
18                                          DNA repair  0.437623  
19                                          DNA repair  0.434165  
20                                          DNA repair  0.433227  
21                                          DNA repair  0.408658  
22                                     DNA replication  0.408610  
23                       interstrand cross-link repair  0.405009  
24                       interstrand cross-link repair  0.404687  
25                                     DNA replication  0.402419  
26                         response to toxic substance  0.401411  
27                      

## Visualizing the output

In addition to parsing the output, we can actually visualize it in robokop.  First, we need to send the quick output to the viewer, and we will return an identifier for the visualization:

In [33]:
view_post_url = f'http://{robokop_server}/api/simple/view/'
view_post_response = requests.post(view_post_url, json=two_step_answer)

In [38]:
from IPython.core.display import display, HTML
uid=view_post_response.json()
view_url = f'http://{robokop_server}/simple/view/{uid}'
display(HTML(f'<a href={view_url}>View Answer</a>'))

## Output Format

### Version 0.9, output_format=message

The results above are returned in v0.9 of the Translator standard for graph answers.  In this format, one graph is returned, containing all of the nodes and edges required for any of the answers.  Then the list of answers is given, binding nodes and edges in the question definition to nodes and edges in the shared knowledge graph.  This allows greater efficiency and specificity in relating outputs to question parameters.

The response has 3 keys returning
1. The json formatted question that was asked 
2. The shared knowledge graph 
3. The answers as defined on that graph

In [29]:
print(two_step_answer.keys())

dict_keys(['question_graph', 'knowledge_graph', 'answers'])


The knowledge graph consists of nodes (containing all the information about the entity) and edges (containing information about the relationship between entities).

In [30]:
print(two_step_answer['knowledge_graph'].keys())

dict_keys(['nodes', 'edges'])


Here's what the first 2 nodes and the first edge look like:

In [31]:
print(json.dumps(two_step_answer['knowledge_graph']['nodes'][:2],indent=4))

[
    {
        "congenital abnormality": true,
        "name": "Fanconi anemia",
        "nutritional or metabolic disease": true,
        "id": "MONDO:0019391",
        "equivalent_identifiers": [
            "MESH:D005199",
            "MEDDRA:10016218",
            "MONDO:0019391",
            "MEDDRA:10055206",
            "UMLS:C0015625",
            "ORPHANET:84",
            "DOID:13636"
        ],
        "type": "disease",
        "rare disease": true,
        "syndromic disease": true
    },
    {
        "name": "UBE2T",
        "location": "1q32.1",
        "locus_group": "protein-coding gene",
        "gene_family": [
            "Ubiquitin conjugating enzymes E2",
            "FA complementation groups"
        ],
        "chromosome": "1",
        "id": "HGNC:25009",
        "equivalent_identifiers": [
            "UniProtKB:Q9NPD8",
            "NCBIGene:29089",
            "HGNC:25009",
            "ENSEMBL:ENSG00000077152",
            "HGNC.SYMBOL:UBE2T",
          

In [33]:
print(json.dumps(two_step_answer['knowledge_graph']['edges'][1],indent=4))

{
    "relation": "PHAROS:gene_involved",
    "target_id": "HGNC:25009",
    "edge_source": "pharos.disease_get_gene",
    "publications": [],
    "id": "4257632",
    "predicate_id": "NCIT:R176",
    "source_database": "pharos",
    "source_id": "MONDO:0019391",
    "type": "disease_to_gene_association",
    "ctime": 1543738834.740142,
    "relation_label": "gene_involved",
    "weight": 0.25832639515782074
}


Here's what the first answer looks like:

In [34]:
print(json.dumps(two_step_answer['answers'][0],indent=4))

{
    "node_bindings": {
        "n0": "MONDO:0019391",
        "n1": "HGNC:20473",
        "n2": "GO:0009636"
    },
    "edge_bindings": {
        "e0": [
            "4257531",
            "6938566",
            "6938600"
        ],
        "e1": [
            "4294578",
            "1094816"
        ]
    },
    "score": 0.3959587237470695
}


The answers are ranked by score, so this is the highest ranking answer, and the score is given.  The `node_bindings` relate the identifiers of the questions (`n0`, `n1`, `n2`) to the identifiers in the knowledge graph (`MONDO:0019391`, `HGNC:20473`, `GO:0009636`) and we could use these identifiers to look up further information about those nodes in the graph.

Similarly, the `e0` edge (connecting `n0` and `n1` in the question), is mapped to several possible edges in the graph, and we could look up the edge information using the identifier.  For instance, we can find the first edge between for `e0`:

In [41]:
scores = [a['score'] for a in two_step_answer['answers']]
scores

[0.3959587237470695,
 0.39562714985581743,
 0.3930911628130927,
 0.38680121546632074,
 0.38680121546632074,
 0.38237448152146924,
 0.37742069307012543,
 0.37742069307012543,
 0.37742069307012543,
 0.35967233858255104,
 0.35967233858255104,
 0.3328828490620355,
 0.3328828490620355,
 0.3328828490620355,
 0.3328828490620355,
 0.3328828490620355,
 0.330896351413593,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.29235569718019727,
 0.22921193146495913,
 0.22921193146495913,
 0.22921193146495913,
 0.22921193146495913,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 0.225403541755474,
 

In [35]:
my_edge = list(filter(lambda x: x['id'] == "4257531", two_step_answer['knowledge_graph']['edges']))[0]
print(json.dumps(my_edge,indent=4))

{
    "relation": "PHAROS:gene_involved",
    "target_id": "HGNC:20473",
    "edge_source": "pharos.disease_get_gene",
    "publications": [],
    "id": "4257531",
    "predicate_id": "NCIT:R176",
    "source_database": "pharos",
    "source_id": "MONDO:0019391",
    "type": "disease_to_gene_association",
    "ctime": 1543738834.73799,
    "relation_label": "gene_involved",
    "weight": 0.25832639515782074
}


### v0.8, output_format=dense

For backwards compatibility, output can also be returned in v0.8 of the KG format.  To return output in this format, pass the argument `output_format=dense` to the url.   The `output_format` defaults to `message` which specifies v0.9.

In [36]:
two_step_answer_dense = quick(two_step_question,output_format='dense')

Return Status: 200


In [38]:
two_step_answer_dense.keys()

dict_keys(['datetime', 'id', 'message', 'response_code', 'result_list'])

In [42]:
two_step_answer_dense['result_list'][0]

{'confidence': 0.3959587237470695,
 'id': '5939821d-f96e-442c-ab37-059ffe1087e4',
 'result_graph': {'edge_list': [{'confidence': 0.25832639515782074,
    'num_publications': None,
    'provided_by': 'pharos.disease_get_gene',
    'publications': '',
    'source_id': 'MONDO:0019391',
    'target_id': 'HGNC:20473',
    'type': 'disease_to_gene_association'},
   {'confidence': 0.4071474314830641,
    'num_publications': None,
    'provided_by': 'biolink.disease_get_gene',
    'publications': 'PMID:20639400',
    'source_id': 'MONDO:0019391',
    'target_id': 'HGNC:20473',
    'type': 'biomarker_for'},
   {'confidence': 0.9998780282217139,
    'num_publications': None,
    'provided_by': 'biolink.disease_get_gene',
    'publications': 'PMID:25980754,PMID:19127258,PMID:16116423,PMID:16153896,PMID:16116424,PMID:27498913,PMID:22006311,PMID:24556621,PMID:26689913,PMID:26681312,PMID:17033622,PMID:26720728,PMID:25186627,PMID:25452441,PMID:26921362,PMID:21964575,PMID:26786923,PMID:21127055,PMID:2

## Specifying multiple fixed nodes

The examples above have all started from one known node and expanded from it, sometimes in multiple steps.  It's also possible to have more than one node specified.  For instance, if we wanted to look for genes that link Fanconi Aneima (MONDO:0019391) and DNA repair (GO:0006281), we could rerun our previous query, but setting the final curie, like this:

In [44]:
two_step_question_fixed_ends = \
   make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,'GO:0006281'])
two_step_answer_fixed_ends = quick(two_step_question_fixed_ends)

Return Status: 200


In [45]:
parse_answer(two_step_answer_fixed_ends,node_list=['n0','n1','n2'],edge_list=[],node_properties=['name'])

n0 - name  n1 - name   n2 - name      score
0   Fanconi Anemia      RAD51  DNA repair  0.452533
1   Fanconi Anemia      FANCC  DNA repair  0.441425
2   Fanconi Anemia     RAD51C  DNA repair  0.440971
3   Fanconi Anemia      FANCG  DNA repair  0.440748
4   Fanconi Anemia      FANCA  DNA repair  0.437623
5   Fanconi Anemia       SLX4  DNA repair  0.434165
6   Fanconi Anemia      FANCL  DNA repair  0.433227
7   Fanconi Anemia      ERCC4  DNA repair  0.408658
8   Fanconi Anemia      XRCC2  DNA repair  0.399073
9   Fanconi Anemia     FANCD2  DNA repair  0.393783
10  Fanconi Anemia      UBE2T  DNA repair  0.385278
11  Fanconi Anemia        XPA  DNA repair  0.327133
12  Fanconi Anemia        BLM  DNA repair  0.323795
13  Fanconi Anemia        WRN  DNA repair  0.322999
14  Fanconi Anemia      MUS81  DNA repair  0.321773
15  Fanconi Anemia     RAD51D  DNA repair  0.320855
16  Fanconi Anemia       FAN1  DNA repair  0.320700
17  Fanconi Anemia   RAD51AP1  DNA repair  0.320261
18  Fanconi Anemia      DDX11  DNA repair  0.320211
19  Fanconi Anemia      CENPS  DNA repair  0.270052
20  Fanconi Anemia      SLX1B  DNA repair  0.270052
21  Fanconi Anemia      SLX1B  DNA repair  0.270052
22  Fanconi Anemia     BOD1L1  DNA repair  0.270052

## Non-path queries

So far, we've only looked at linear paths.  But the question format is actually more general than that - we can define a path pattern generally. So for instance, in the above query, we find all genes that are linked to both FA and DNA repair.  But what if we wanted to find entities that are connected to more than two specified entities.  Here is a query-generation function for the star query:

In [16]:
def make_star_question(types,curies,shared_type,rebuild=None):
    """Create a question to find entities of shared_type that are linked to all of the nodes specified in the
    types and curies arrays."""
    question = {
                #'rebuild': 'True',
                'machine_question': {
                    'nodes': [],
                    'edges': []
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild']='True'
    question['machine_question']['nodes'].append( {'id': 'n0', 'type': shared_type})
    for i,t in enumerate(types):
        newnode = {'id': f'n{i+1}', 'type': t}
        if curies[i] is not None:
            newnode['curie'] = curies[i]
        question['machine_question']['nodes'].append(newnode)
        question['machine_question']['edges'].append( {'id': f'e{i}', 'source_id': 'n0', 'target_id': f'n{i+1}'})
    return question

Suppose I have this set of GO terms, and I'd like to find genes that they all have in common:

* 'voltage-gated sodium channel activity(GO:0005248)',
* 'muscle contraction(GO:0006936)',
* 'voltage-gated ion channel activity(GO:0005244)',
* 'regulation of ion transmembrane transport(GO:0034765)',
* 'sodium ion transmembrane transport(GO:0035725)',
* 'neuronal action potential(GO:0019228)',
* 'membrane depolarization during action potential(GO:0086010)',
* 'sodium ion transport(GO:0006814)'

In [17]:
go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814']
types = ['biological_process_or_activity' for g in go_terms]
star_q = make_star_question(types,go_terms,'gene')
star_q

{'machine_question': {'edges': [{'id': 'e0',
    'source_id': 'n0',
    'target_id': 'n1'},
   {'id': 'e1', 'source_id': 'n0', 'target_id': 'n2'},
   {'id': 'e2', 'source_id': 'n0', 'target_id': 'n3'},
   {'id': 'e3', 'source_id': 'n0', 'target_id': 'n4'},
   {'id': 'e4', 'source_id': 'n0', 'target_id': 'n5'},
   {'id': 'e5', 'source_id': 'n0', 'target_id': 'n6'},
   {'id': 'e6', 'source_id': 'n0', 'target_id': 'n7'},
   {'id': 'e7', 'source_id': 'n0', 'target_id': 'n8'}],
  'nodes': [{'id': 'n0', 'type': 'gene'},
   {'curie': 'GO:0005248',
    'id': 'n1',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0006936',
    'id': 'n2',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0005244',
    'id': 'n3',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0034765',
    'id': 'n4',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0035725',
    'id': 'n5',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0019228',
   

In [18]:
common_gene_answer = quick(star_q)

Return Status: 200


In [19]:
extract_node_names(common_gene_answer)

n0          n1                  n2                                  n3  \
0  SCN4A  GO:0005248  muscle contraction  voltage-gated ion channel activity   
1  SCN7A  GO:0005248  muscle contraction  voltage-gated ion channel activity   

                                          n4  \
0  regulation of ion transmembrane transport   
1  regulation of ion transmembrane transport   

                                   n5                         n6  \
0  sodium ion transmembrane transport  neuronal action potential   
1  sodium ion transmembrane transport  neuronal action potential   

                                                n7                    n8  
0  membrane depolarization during action potential  sodium ion transport  
1  membrane depolarization during action potential  sodium ion transport

We can also have more than one unspecified node.  Suppose, for instance that we wanted to do the previous query, but we also wanted to know what chemicals interact with the genes that we find.  We can do another star query, where one of our spokes is unspecified:

In [20]:
go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814',None]
types = ['biological_process_or_activity' for i in range(8)]+['chemical_substance']
star_q_compound = make_star_question(types,go_terms,'gene',rebuild=True)
star_q_compound

{'machine_question': {'edges': [{'id': 'e0',
    'source_id': 'n0',
    'target_id': 'n1'},
   {'id': 'e1', 'source_id': 'n0', 'target_id': 'n2'},
   {'id': 'e2', 'source_id': 'n0', 'target_id': 'n3'},
   {'id': 'e3', 'source_id': 'n0', 'target_id': 'n4'},
   {'id': 'e4', 'source_id': 'n0', 'target_id': 'n5'},
   {'id': 'e5', 'source_id': 'n0', 'target_id': 'n6'},
   {'id': 'e6', 'source_id': 'n0', 'target_id': 'n7'},
   {'id': 'e7', 'source_id': 'n0', 'target_id': 'n8'},
   {'id': 'e8', 'source_id': 'n0', 'target_id': 'n9'}],
  'nodes': [{'id': 'n0', 'type': 'gene'},
   {'curie': 'GO:0005248',
    'id': 'n1',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0006936',
    'id': 'n2',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0005244',
    'id': 'n3',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0034765',
    'id': 'n4',
    'type': 'biological_process_or_activity'},
   {'curie': 'GO:0035725',
    'id': 'n5',
    'type': 'biologica

**WARNING: This takes a long time to run**

In [23]:
common_gene_compound_answer = quick(star_q_compound)

Return Status: 200


In [24]:
cgc_nodes = extract_node_names(common_gene_compound_answer)
cgc_nodes[['n0','n9']]

n0                                                 n9
0    SCN4A                                           m-cresol
1    SCN4A                        4-chloro-3,5-dimethylphenol
2    SCN4A                                     4-chlorophenol
3    SCN7A                                           ethotoin
4    SCN7A                                         benzocaine
5    SCN7A                                          tocainide
6    SCN4A                                       pilsicainide
7    SCN4A                                        lamotrigine
8    SCN4A                                           ethotoin
9    SCN7A                                    phenazopyridine
10   SCN4A                         procainamide hydrochloride
11   SCN4A                                          phenytoin
12   SCN7A                         procainamide hydrochloride
13   SCN4A                                         flecainide
14   SCN4A          levobupivacaine hydrochloride (anhydrous)
15   SCN4A                                    (S)-ropivacaine
16   SCN4A                                          lidocaine
17   SCN4A                                         etidocaine
18   SCN4A                                        fospropofol
19   SCN4A                                         benzocaine
20   SCN4A  1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...
21   SCN4A                                          tocainide
22   SCN4A                                         zonisamide
23   SCN7A                                         rufinamide
24   SCN4A                                    phenazopyridine
25   SCN4A                                        mepivacaine
26   SCN7A                                        mephenytoin
27   SCN7A                                         sodium(1+)
28   SCN4A                                         mexiletine
29   SCN4A  [[4-(4-fluorophenoxy)phenyl]methylideneamino]urea
..     ...                                                ...
140  SCN4A                            eslicarbazepine acetate
141  SCN4A         (S)-2-(4-chlorophenoxy)-1-phenylethanamine
142  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine
143  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine
144  SCN4A  (S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...
145  SCN4A                                                 NA
146  SCN4A                                                 NA
147  SCN4A  N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...
148  SCN4A                                                 NA
149  SCN4A                                                 NA
150  SCN4A                                                 NA
151  SCN4A                                                 NA
152  SCN4A                                                 NA
153  SCN4A                                                 NA
154  SCN4A                                                 NA
155  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...
156  SCN4A                                       orphenadrine
157  SCN4A                                                 NA
158  SCN4A                                                 NA
159  SCN4A     (R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine
160  SCN4A                                                 NA
161  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole
162  SCN4A                                                 NA
163  SCN4A                                            HWTX-IV
164  SCN4A                                                 NA
165  SCN4A                                                 NA
166  SCN4A                                                 NA
167  SCN4A                                                 NA
168  SCN4A                                                 NA
169  SCN4A                                                 NA

[170 rows x 2 columns]

Many other graph shapes are possible, these are simply a couple of examples.  For instance, we could add onto this query an edge from node 0 to a new disease node (node 10) and also include an edge from the target node (node 9) to this same disease node. Now we would be finding genes that share a set of GO terms, then looking for disease,drug pairs such that the disease, the drug, and the gene are all interconnected.

### Controlling the number of results

By default, quick will return up to 250 results, but this can be raised or lowered using the max_results parameter.  We've already rebuilt this question, so we won't do rebuild=True this time, which should make it go faster.

In [27]:
star_q_compound = make_star_question(types,go_terms,'gene')
common_gene_compound_answer_10 = quick(star_q_compound,max_results=10)

Return Status: 200


In [28]:
cgc_nodes_10 = extract_node_names(common_gene_compound_answer_10)
cgc_nodes_10[['n0','n9']]

n0                                                 n9
0    SCN4A                                           m-cresol
1    SCN4A                        4-chloro-3,5-dimethylphenol
2    SCN4A                                     4-chlorophenol
3    SCN7A                                           ethotoin
4    SCN7A                                         benzocaine
5    SCN7A                                          tocainide
6    SCN4A                                       pilsicainide
7    SCN4A                                        lamotrigine
8    SCN4A                                           ethotoin
9    SCN7A                                    phenazopyridine
10   SCN4A                         procainamide hydrochloride
11   SCN4A                                          phenytoin
12   SCN7A                         procainamide hydrochloride
13   SCN4A                                         flecainide
14   SCN4A          levobupivacaine hydrochloride (anhydrous)
15   SCN4A                                    (S)-ropivacaine
16   SCN4A                                          lidocaine
17   SCN4A                                         etidocaine
18   SCN4A                                        fospropofol
19   SCN4A                                         benzocaine
20   SCN4A  1-butyl-N-(2,6-dimethylphenyl)piperidine-2-car...
21   SCN4A                                          tocainide
22   SCN4A                                         zonisamide
23   SCN7A                                         rufinamide
24   SCN4A                                    phenazopyridine
25   SCN4A                                        mepivacaine
26   SCN7A                                        mephenytoin
27   SCN7A                                         sodium(1+)
28   SCN4A                                         mexiletine
29   SCN4A  [[4-(4-fluorophenoxy)phenyl]methylideneamino]urea
..     ...                                                ...
140  SCN4A                            eslicarbazepine acetate
141  SCN4A         (S)-2-(4-chlorophenoxy)-1-phenylethanamine
142  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine
143  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine
144  SCN4A  (S)-1-benzyl-N-(2,6-dimethylphenyl)pyrrolidine...
145  SCN4A                                                 NA
146  SCN4A                                                 NA
147  SCN4A  N-(2-Methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...
148  SCN4A                                                 NA
149  SCN4A                                                 NA
150  SCN4A                                                 NA
151  SCN4A                                                 NA
152  SCN4A                                                 NA
153  SCN4A                                                 NA
154  SCN4A                                                 NA
155  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...
156  SCN4A                                       orphenadrine
157  SCN4A                                                 NA
158  SCN4A                                                 NA
159  SCN4A     (R)-2-(2,6-dimethylphenoxy)-1-phenylethanamine
160  SCN4A                                                 NA
161  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole
162  SCN4A                                                 NA
163  SCN4A                                            HWTX-IV
164  SCN4A                                                 NA
165  SCN4A                                                 NA
166  SCN4A                                                 NA
167  SCN4A                                                 NA
168  SCN4A                                                 NA
169  SCN4A                                                 NA

[170 rows x 2 columns]

## Controlling the generality and speed with max_connectivity

ROBOKOP has access to nodes that range from very specific, to highly general. For instance, an anatomical_feature may represent a dopaminergic_neuron, or a neuron, or the midbrain, or the brain or the nervous system.  A biological process may be drawn from anywhere in the very large GO ontology.

This has two effects: first, general nodes are highly connected, since e.g. the nervous system is involved in many diseases, where a specfic part of the nervous system may be involved in only very specfic diseases.  Second, because these nodes have many relationships, many paths go through these nodes, and they make the run time of queries much longer.

This can be controlled for with the `max_connectivity` parameter, which is the maximum degree of an allowed node in a returned answer.   So if the parameter is set to 1000, then no node with more than 1000 edges in the database will be included in path finding.  The largely has the effect of removing very general nodes, but this is not guaranteed.

In [12]:
two_step_question = make_N_step_question(['disease','biological_process_or_activity','gene'],['MONDO:0019391',None,None])

In [20]:
all_answers = quick(two_step_question,max_results=10)

Return Status: 200


In [21]:
len(all_answers['answers'])

250

## Rebuild and caching

For details on ROBOKOP caching, and guidance for when to use the `rebuild` parameter, see the notebook on the expand service.  Briefly, the quick service will only run against the cache unless the rebuild parameter is passed, in which case services are re-queried.   The rebuild parameter is passed as part of the json query as seen in the `make_star_query` function above. 

## Edge Properties

In the expand notebook, we saw how a particular predicate could be chosen to limit the results.  This same type information can in fact be added to any edge in a quick query, providing fine control to the results returned.